arbitrator/prover/src/machine.rs

```diff
@@ -2483,9 +2517,34 @@ impl Machine {
                 hash_stack_with_heights(frames, &heights, concat!($prefix, " stack:"))
             }};
         }
+
+        // compute_slices works similarly as compute, except for vector of slices.
+        // Instead of constructing vector of each Value's hash, it constructs
+        // hash of each slice (&[Value]). Rest is the same.
+        macro_rules! compute_vec {
+            ($field:expr, $stacks:expr, $prefix:expr) => {{
+                let heights: Vec<_> = self.guards.iter().map($field).collect();
+                
+                // Map each &[Value] slice to its hash using `hash_stack` and collect these hashes into a vector
+                // Reorder stacks for hashing to follow same order as in proof:
+                // [0, n-1, 1, 2, ..., n - 2].
+                let ordered_stacks: Vec<_> = match $stacks.len() {
+                    0 => Vec::new(),
+                    1 => vec![*$stacks.first().unwrap()],
+                    _ => std::iter::once(*$stacks.first().unwrap())
+                        .chain(std::iter::once(*$stacks.last().unwrap()))
+                        .chain($stacks.iter().skip(1).take($stacks.len() - 2).cloned())
+                        .collect()
+                };
+                let slice_hashes: Vec<_> = ordered_stacks.iter()
+                    .map(|slice| hash_stack(slice.iter().map(|v| v.hash()), concat!($prefix, " stack:")))
+                    .collect();
+                hash_stack_with_heights(slice_hashes, &heights, concat!($prefix, " stack:"))
+            }};
+        }
         let (frame_stack, frames) =
-            compute!(|x| x.frame_stack, self.get_frame_stack(), "Stack frame");
-        let (value_stack, values) = compute!(|x| x.value_stack, self.get_data_stack(), "Value");
+            compute_vec!(|x| x.frame_stack, self.get_frame_stacks(), "Stack frame");
+        let (value_stack, values) = compute_vec!(|x| x.value_stack, self.get_data_stacks(), "Value");
         let (inter_stack, inters) = compute!(|x| x.inter_stack, self.internal_stack, "Value");
```

Comment:
For each multistack, we'll return a multistack hash, computed like this:
```
multistack_hash = keccak("multistack: " + main_stack + last_cothread + other_cothreads)
where:
last_cothread = bytes32(0x0) if no cothreads stack_hash otherwise
other_cothread=
if more than one cothread: keccak("cothread: " + cothread[n-2] + keccak("cothread: " + cothread[n-3] + keccak(... + keccak("cothread: "  + cothread[1])
else: bytes32(0x0)
```

frames, values from above are for error guards - they will only be taken from the main stack (stack 0)

Reply:
changed, ptal.